File: docs/src/test/java/docs/javadsl/CreateTablesDocExample.java
/*
 * Copyright (C) 2024 Lightbend Inc. <https://www.lightbend.com>
 */

package docs.javadsl;

import java.util.concurrent.TimeUnit;

import akka.actor.typed.ActorSystem;

// #create-tables
import akka.persistence.dynamodb.DynamoDBSettings;
import akka.persistence.dynamodb.util.ClientProvider;
import akka.persistence.dynamodb.util.javadsl.CreateTables;
import software.amazon.awssdk.services.dynamodb.DynamoDbAsyncClient;

// #create-tables

public class CreateTablesDocExample {

  public static void createTables(ActorSystem<?> system) throws Exception {
    // #create-tables
    String dynamoDBConfigPath = "akka.persistence.dynamodb";
    String dynamoDBClientConfigPath = dynamoDBConfigPath + ".client";

    DynamoDBSettings settings =
        DynamoDBSettings.create(system.settings().config().getConfig(dynamoDBConfigPath));

    DynamoDbAsyncClient client = ClientProvider.get(system).clientFor(dynamoDBClientConfigPath);

    // create journal table, synchronously
    CreateTables.createJournalTable(system, settings, client, /*deleteIfExists:*/ true)
        .toCompletableFuture()
        .get(10, TimeUnit.SECONDS);

    // create snapshot table, synchronously
    CreateTables.createSnapshotsTable(system, settings, client, /*deleteIfExists:*/ true)
        .toCompletableFuture()
        .get(10, TimeUnit.SECONDS);
    // #create-tables
  }

}

File: docs/src/test/scala/docs/scaladsl/CreateTablesDocExample.scala
/*
 * Copyright (C) 2024 Lightbend Inc. <https://www.lightbend.com>
 */

package docs.scaladsl

import scala.concurrent.Await
import scala.concurrent.duration._

import akka.actor.testkit.typed.scaladsl.ScalaTestWithActorTestKit
import com.typesafe.config.Config
import com.typesafe.config.ConfigFactory
import org.scalatest.concurrent.Futures
import org.scalatest.wordspec.AnyWordSpecLike

//#create-tables
import akka.persistence.dynamodb.DynamoDBSettings
import akka.persistence.dynamodb.util.ClientProvider
import akka.persistence.dynamodb.util.scaladsl.CreateTables
import software.amazon.awssdk.services.dynamodb.DynamoDbAsyncClient

//#create-tables

object CreateTablesDocExample {
  val config: Config = ConfigFactory.load(ConfigFactory.parseString("""
    //#local-mode
    akka.persistence.dynamodb {
      client.local.enabled = true
    }
    //#local-mode
    """))
}

class CreateTablesDocExample
    extends ScalaTestWithActorTestKit(CreateTablesDocExample.config)
    with AnyWordSpecLike
    with Futures {

  "Getting started docs" should {

    "have example of creating tables locally (Scala)" in {
      //#create-tables
      val dynamoDBConfigPath = "akka.persistence.dynamodb"
      val dynamoDBClientConfigPath = dynamoDBConfigPath + ".client"

      val settings: DynamoDBSettings = DynamoDBSettings(system.settings.config.getConfig(dynamoDBConfigPath))
      val client: DynamoDbAsyncClient = ClientProvider(system).clientFor(dynamoDBClientConfigPath)

      // create journal table, synchronously
      Await.result(CreateTables.createJournalTable(system, settings, client, deleteIfExists = true), 10.seconds)

      // create snapshot table, synchronously
      Await.result(CreateTables.createSnapshotsTable(system, settings, client, deleteIfExists = true), 10.seconds)
      //#create-tables
    }

    "have example of creating tables locally (Java)" in {
      docs.javadsl.CreateTablesDocExample.createTables(system)
    }

  }
}
